### ISSUE TITLE
Gravity acceleration removal

### ISSUE THREAD

**Issue by ricber** [xioTechnologies/Fusion#33]
Hi, thank you for your work! I was wondering if your code removes the gravity acceleration. It seems that it does it, but I'm not sure how. Also, does it need magnetometer data to remove the gravity, or is it sufficient to have accelerometer and gyroscope data? Because I tried the code with my collected data and the gravity was not removed.

**Comment by xioTechnologies** [comment]
From the [README](https://github.com/xioTechnologies/Fusion/blob/main/README.md):
"_The algorithm provides three outputs: quaternion, linear acceleration, and Earth acceleration. ... The linear acceleration is the accelerometer measurement with the 1 g of gravity removed. The Earth acceleration is the accelerometer measurement in the Earth coordinate frame with the 1 g of gravity removed._"

A magnetometer is not required for the calculation of linear acceleration or Earth acceleration.

**Comment by ricber** [comment]
Does the gravity removal start from the first message or only after an initialization phase? Also, does it require the IMU to be calibrated?

**Comment by xioTechnologies** [comment]
You should assume that the algorithm outputs are inaccurate until initialisation is complete.  The accuracy of algorithm outputs will depend on the accuracy of the sensor measurements you provide.  [Garbage in, garbage out](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out).

**Comment by ricber** [comment]
Hi! I tried your filter, but the gravity removal doesn't seem to be happening. Do you know which are possible causes of this behavior?

**Comment by xioTechnologies** [comment]
There are no know issues with `earth_acceleration` or `linear_acceleration`.  Please can you provide a description of the issue,

**Comment by ricber** [comment]
Hi, I am trying to do dead reckoning by using an adaptation of your `gait_tracking.py` script. I removed the detection of the moving periods and used the `linear_acceleration` instead of the `earth_acceleration`. For my use case, it is important to know the transformation between successive positions accurately and not have an accurate long-run positioning. So I can ignore the drift. I run the script with my IMU data, but the plot of the `linear_acceleration` reveals that gravity has not been removed on the Z component. 

My script:
```python
from matplotlib import animation
import imufusion
import matplotlib.pyplot as pyplot
import numpy

# Import sensor data ("short_walk.csv" or "long_walk.csv")
data = numpy.genfromtxt("my_walk.csv", delimiter=",", skip_header=1)

sample_rate = 110  # Hz

timestamp = data[:, 0]
gyroscope = data[:, 1:4]
accelerometer = data[:, 4:7]

# Plot sensor data
figure, axes = pyplot.subplots(nrows=6, sharex=True, gridspec_kw={"height_ratios": [6, 6, 6, 2, 1, 1]})

figure.suptitle("Sensors data, Euler angles, and AHRS internal states")

axes[0].plot(timestamp, gyroscope[:, 0], "tab:red", label="Gyroscope X")
axes[0].plot(timestamp, gyroscope[:, 1], "tab:green", label="Gyroscope Y")
axes[0].plot(timestamp, gyroscope[:, 2], "tab:blue", label="Gyroscope Z")
axes[0].set_ylabel("Degrees/s")
axes[0].grid()
axes[0].legend()

axes[1].plot(timestamp, accelerometer[:, 0], "tab:red", label="Accelerometer X")
axes[1].plot(timestamp, accelerometer[:, 1], "tab:green", label="Accelerometer Y")
axes[1].plot(timestamp, accelerometer[:, 2], "tab:blue", label="Accelerometer Z")
axes[1].set_ylabel("g")
axes[1].grid()
axes[1].legend()

# Intantiate AHRS algorithms
offset = imufusion.Offset(sample_rate)
ahrs = imufusion.Ahrs()

ahrs.settings = imufusion.Settings(0.5,  # gain
                                   10,  # acceleration rejection
                                   0,  # magnetic rejection
                                   5 * sample_rate)  # rejection timeout = 5 seconds

# Process sensor data
delta_time = numpy.diff(timestamp, prepend=timestamp[0])

euler = numpy.empty((len(timestamp), 3))
internal_states = numpy.empty((len(timestamp), 3))
acceleration = numpy.empty((len(timestamp), 3))

for index in range(len(timestamp)):
    gyroscope[index] = offset.update(gyroscope[index])

    ahrs.update_no_magnetometer(gyroscope[index], accelerometer[index], delta_time[index])

    euler[index] = ahrs.quaternion.to_euler()

    ahrs_internal_states = ahrs.internal_states
    internal_states[index] = numpy.array([ahrs_internal_states.acceleration_error,
                                          ahrs_internal_states.accelerometer_ignored,
                                          ahrs_internal_states.acceleration_rejection_timer])

    acceleration[index] = ahrs.linear_acceleration

# Plot Euler angles
axes[2].plot(timestamp, euler[:, 0], "tab:red", label="Roll")
axes[2].plot(timestamp, euler[:, 1], "tab:green", label="Pitch")
axes[2].plot(timestamp, euler[:, 2], "tab:blue", label="Yaw")
axes[2].set_ylabel("Degrees")
axes[2].grid()
axes[2].legend()

# Plot internal states
axes[3].plot(timestamp, internal_states[:, 0], "tab:olive", label="Acceleration error")
axes[3].set_ylabel("Degrees")
axes[3].grid()
axes[3].legend()

axes[4].plot(timestamp, internal_states[:, 1], "tab:cyan", label="Accelerometer ignored")
pyplot.sca(axes[4])
pyplot.yticks([0, 1], ["False", "True"])
axes[4].grid()
axes[4].legend()

axes[5].plot(timestamp, internal_states[:, 2], "tab:orange", label="Acceleration rejection timer")
axes[5].set_xlabel("Seconds")
axes[5].grid()
axes[5].legend()

# Plot acceleration
_, axes = pyplot.subplots(nrows=4, sharex=True, gridspec_kw={"height_ratios": [6, 1, 6, 6]})

axes[0].plot(timestamp, acceleration[:, 0], "tab:red", label="X")
axes[0].plot(timestamp, acceleration[:, 1], "tab:green", label="Y")
axes[0].plot(timestamp, acceleration[:, 2], "tab:blue", label="Z")
axes[0].set_title("Acceleration")
axes[0].set_ylabel("m/s/s")
axes[0].grid()
axes[0].legend()


# Calculate velocity (includes integral drift)
velocity = numpy.zeros((len(timestamp), 3))

for index in range(len(timestamp)):
    velocity[index] = velocity[index - 1] + delta_time[index] * acceleration[index]


# Plot velocity
axes[2].plot(timestamp, velocity[:, 0], "tab:red", label="X")
axes[2].plot(timestamp, velocity[:, 1], "tab:green", label="Y")
axes[2].plot(timestamp, velocity[:, 2], "tab:blue", label="Z")
axes[2].set_title("Velocity")
axes[2].set_ylabel("m/s")
axes[2].grid()
axes[2].legend()

# Calculate position
position = numpy.zeros((len(timestamp), 3))

for index in range(len(timestamp)):
    position[index] = position[index - 1] + delta_time[index] * velocity[index]

# Plot position
axes[3].plot(timestamp, position[:, 0], "tab:red", label="X")
axes[3].plot(timestamp, position[:, 1], "tab:green", label="Y")
axes[3].plot(timestamp, position[:, 2], "tab:blue", label="Z")
axes[3].set_title("Position")
axes[3].set_xlabel("Seconds")
axes[3].set_ylabel("m")
axes[3].grid()
axes[3].legend()

pyplot.show()
```

My data: 
[my_walk.csv](https://github.com/xioTechnologies/Fusion/files/9001353/my_walk.csv)

Plots:
![Figure_1](https://user-images.githubusercontent.com/18362950/176190555-b9348a8e-d8e0-4217-9f01-16e95683dc7a.png)
![Figure_2](https://user-images.githubusercontent.com/18362950/176190580-02cb8b90-4c7c-46a5-a875-1d404e65b504.png)

**Comment by xioTechnologies** [comment]
The problem is that you are providing accelerometer measurements in m/s/s, not g.  The algorithm expects accelerometer measurements in g as described by the documentation, `* @param accelerometer Accelerometer measurement in g.` on [line 96](https://github.com/xioTechnologies/Fusion/blob/main/Fusion/FusionAhrs.c#L96).

In addition to this documentation:
- You modified the [example code](https://github.com/xioTechnologies/Gait-Tracking/blob/main/gait_tracking.py#L64) to replace `acceleration[index] = 9.81 * ahrs.earth_acceleration  # convert g to m/s/s` with `acceleration[index] = ahrs.linear_acceleration`, deleting a comment that clarifies the units of measurement.
- You opened a [separate issue](https://github.com/xioTechnologies/Gait-Tracking/issues/7) to ask about this line of code where I clarified that the units are g and not m/s/s.
- Your plot above shows the accelerometer is measuring 10 g while stationary - a obvious hint that you are using m/s/s when g is expected.

You also need to modify your code above to use `earth_acceleration` as per the example code, not `linear_acceleration`.

**Comment by ricber** [comment]
@xioTechnologies I see, so I need to divide my accelerometer data by 9.81. Thank you! I removed the comment from the code because I thought the `linear_acceleration` was in m/s/s. As far as I have understood, the earth acceleration is relative to the Earth coordinate frame, while the liner acceleration is relative to what?

**Comment by xioTechnologies** [comment]
`linear_acceleration` and `earth_acceleration` are in g.  The documentation has been updated to clarify this.

Linear acceleration is relative to whatever your accelerometer is relative to.  _"The linear acceleration is the accelerometer measurement with the 1 g of gravity removed."_  This is a quote from the [README](https://github.com/xioTechnologies/Fusion#readme) and something I already shared above.

**Comment by ricber** [comment]
Thank you for being so helpful! I think it is a good idea to update the documentation. I also suggest updating the [README](https://github.com/xioTechnologies/Fusion#readme) to explain better that by **g**, you mean a unit of measurement. Since **g** is not part of the [SI units](https://en.wikipedia.org/wiki/International_System_of_Units), it is not immediate to get its meaning for a non-expert of the subject. Moreover, when you answered my [separate issue](https://github.com/xioTechnologies/Gait-Tracking/issues/7), you attached to [g](https://en.wikipedia.org/wiki/Gravity_of_Earth) a link pointing to the Wikipedia page [Gravity of Earth](https://en.wikipedia.org/wiki/Gravity_of_Earth). As you can read from that page:

> The gravity of Earth, denoted by **g**, is the net acceleration that is imparted to objects due to the combined effect of gravitation (from mass distribution within Earth) and the centrifugal force (from the Earth's rotation). [...] In SI units this acceleration is expressed in **metres per second squared** [...]

As you can notice, **g** is defined as an acceleration, not as a unit of measurement.

**Comment by Sinou12** [comment]
Hello, 
thank you for your feedback. 
I know that this code is not made for this kind of movement. 
But from your experience, I would like to know your opinion.

I have done a 20 cm displacement along the Y axis only, and I have guaranteed that there will be no rotation along the 3 axes.
The gyroscope displays 0 deg/s in all 3 axes. 

To extract the gravity from the acceleration to calculate the displacement along the Y axis.  Do I just need a filter to extract it? 
Or according to the Y axis, I won't need to do it, because the accelerometer displays 1 g according to Z only?
Since the rotation speed is equal to 0 deg/s, so the rotation matrix is zero, so I don't need to calculate the acceleration in the terrestrial frame (I mean it's the same in both frames)? 

Thank you.

**Comment by xioTechnologies** [comment]
@ricber I do not recognise your issue with g.  The use of g as a unit of measurement is common practise.  It is even used in common parlance.  For example, when talking about rolla coasters, motor sports, aeroplanes, etc.  Most importantly, g is the conventional unit of measurement for accelerometers.  If you check the datasheet of the device you are using, I expect you will see g being used, and m/s/s not mentioned at all.

SI units are not a useful indication of convention.  Kelvin is SI but °C is conventional, radians is SI but degrees is conventional, V/A is SI but Ω is conventional, and so on.

**Comment by ricber** [comment]
@xioTechnologies thank you for the feedback! I understand your point but in ROS, m/s/s is the standard unit of measure for the [IMU message](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/Imu.html). This is why I thought that the convention is to use m/s/s.

**Comment by Taeyoung96** [comment]
@ricber @xioTechnologies 
Hi maybe like you, I am going to use ROS IMU message and use dead reconning with the Fusion algorithm.  
To change the acceleration unit, I did the following process, but it didn't work much.  
Below is part of the source code I used.  
```cpp
const sensor_msgs::Imu thisImu = *imu_msg;
float delta_time = (thisImu.header.stamp - estimate_timestamp_).toSec();    // unit : s

auto acc = thisImu.linear_acceleration;  // unit : m/s^2
FusionVector accelerometer = {float(acc.x / 9.805f) , float(acc.y / 9.805f), float(acc.z / 9.805f)}; 

auto gyr = thisImu.angular_velocity;  // unit : rad/s
FusionVector gyroscope = {rad2deg(gyr.x), rad2deg(gyr.y), rad2deg(gyr.z)}; 

FusionAhrsUpdateNoMagnetometer(&ahrs, gyroscope, accelerometer, delta_time);  

FusionEuler euler = FusionQuaternionToEuler(FusionAhrsGetQuaternion(&ahrs));
cout << "roll :" << euler.angle.roll << " pitch: " << euler.angle.pitch << " yaw: " << euler.angle.yaw << endl;  

FusionVector earth = FusionAhrsGetEarthAcceleration(&ahrs); // unit : g
cout << "earth (g): " << earth.axis.x << " " << earth.axis.y << " " << earth.axis.z << endl;
                
earth = FusionVectorMultiplyScalar(earth, 9.81f); // unit : m/s^2
cout << "earth (m/s^2): " << earth.axis.x << " " << earth.axis.y << " " << earth.axis.z << endl;
```  

And the result looks like below.  
While the roll, pitch, and yaw were well estimated in static conditions,  
the acceleration does not appear to have eliminated the 1g gravity well.  

```
roll :1.42263 pitch: 1.63267 yaw: 1.6682e-09
earth (g): 0.0252494 0.0511634 0.0576138
earth (m/s^2): 0.247697 0.501913 0.565192
```

And after calculating the position, I found that it kept diverging in the z-axis direction.  

It feels similar to the problem that @ricber occured, how did you solve the problem?
